File: 프로그래머스/2/17679. ［1차］ 프렌즈4블록/［1차］ 프렌즈4블록.js
function findRemoveBlock(board){
    const removeArea = new Set();
    for(let i = 0 ; i < board.length - 1; i++){
        for(let j = 0 ; j < board[0].length - 1; j++){
            const cur = board[i][j];
            if (!cur) continue;
            if (cur === board[i][j + 1] && cur === board[i + 1][j] && cur === board[i + 1][j + 1]) {
                removeArea.add(`${i},${j}`);
                removeArea.add(`${i},${j + 1}`);
                removeArea.add(`${i + 1},${j}`);
                removeArea.add(`${i + 1},${j + 1}`);
            }
        }
    }
    return Array.from(removeArea).map(pos => pos.split(',').map(Number));
}

function removeBlock(board, positions){
    let cnt = 0;
    for(const [rx, ry] of positions){
        if(board[rx][ry]){
            board[rx][ry] = "";
            cnt++;
        }
    }
    return cnt;
}

function applyRemoveState(board){
    const row = board.length;
    const col = board[0].length;
    
    for (let j = 0; j < col; j++) {
        const stack = [];

        for (let i = row - 1; i >= 0; i--) {
            if (board[i][j] !== "") stack.push(board[i][j]);
        }

        for (let i = row - 1; i >= 0; i--) {
            board[i][j] = stack[row - 1 - i] || "";
        }
    }
}

function solution(m, n, board) {
    const arrBoard = board.map(row => row.split(""));
    let result = 0;
    
    while(true){
        const goRemove = findRemoveBlock(arrBoard);
        if(goRemove.length === 0) break;
        
        const removed = removeBlock(arrBoard, goRemove);
        result += removed;
        
        applyRemoveState(arrBoard);
    }
    
    return result;
}

File: 프로그래머스/2/17679. ［1차］ 프렌즈4블록/［1차］ 프렌즈4블록.py
def find_remove_block(m, n, board):
    remove_area = set()
    
    for i in range(m-1):
        for j in range(n-1):
            cur = board[i][j]
            if not cur:
                continue
            if cur == board[i][j+1] == board[i+1][j] == board[i+1][j+1]:
                remove_area.update([(i, j), (i, j+1), (i+1, j), (i+1, j+1)])    
    return list(remove_area)

def remove_block(board, positions):
    cnt = 0
    for rx, ry in positions:
        if board[rx][ry]:
            board[rx][ry] = ""
            cnt += 1
    return cnt

def apply_remove_state(m, n, board):
    for j in range(n):
        stack = []
        for i in reversed(range(m)):
            if board[i][j]:
                stack.append(board[i][j])
        for i in reversed(range(m)):
            board[i][j] = stack[m - 1 - i] if m - 1 - i < len(stack) else ""
            
def solution(m, n, board):
    arr_board = [list(row) for row in board]
    result = 0
    
    while True:
        go_remove = find_remove_block(m, n, arr_board)
        if not go_remove: 
            break
        
        removed = remove_block(arr_board, go_remove)
        result += removed
        
        apply_remove_state(m, n, arr_board)
    return result
